File: python/CNTtools/tools/laplacian.py
import numpy as np
from scipy.spatial.distance import cdist
from beartype import beartype
from .pseudo_laplacian import pseudo_laplacian
from numbers import Number


@beartype
def laplacian(data: np.ndarray, labels: np.ndarray, locs: np.ndarray, radius: Number):
    """
    Apply Laplacian re-referencing to multi-channel iEEG data.

    Parameters:
        values (np.ndarray): Matrix representing iEEG data of shape samples X channels.
        labels (np.ndarray): List of channel labels corresponding to the columns of the data matrix.
        locs (np.ndarray): Matrix representing the locations of electrodes.
        radius (float): Radius defining the neighborhood for Laplacian re-referencing.

    Returns:
        np.ndarray: Laplacian-referenced EEG data matrix.
        np.ndarray: Channel labels for the Laplacian-referenced data.
    """
    out_values = np.nan * np.zeros(data.shape)
    nchs = data.shape[1]
    laplacian_labels = []

    nan_elecs = np.any(np.isnan(locs), axis=1)
    pseudo_values, pseudo_labels = pseudo_laplacian(data, labels)

    # Calculate pairwise distances between X and Y
    D = cdist(locs, locs)

    close = D < radius
    close[np.eye(close.shape[0]).astype(bool)] = 0

    # close_chs = []

    for i in range(nchs):
        if not nan_elecs[i]:
            close_elecs = np.nonzero(close[i, :])[0]
            if len(close_elecs) > 0:
                out_values[:, i] = data[:, i] - np.nanmean(data[:, close_elecs], 1)
                laplacian_labels.append(labels[i])
                # close_chs.append(close_elecs)
            else:
                out_values[:, i] = pseudo_values[:, i]
                laplacian_labels.append(pseudo_labels[i])
        else:
            out_values[:, i] = pseudo_values[:, i]
            laplacian_labels.append(pseudo_labels[i])

    return out_values, laplacian_labels

File: matlab/tools/references/laplacian.m
function [out_values,laplacian_labels] = laplacian(values,labels,locs,radius)
% Apply Laplacian re-referencing to multi-channel iEEG data.
%
% Parameters:
% - values (numeric array): Matrix representing iEEG data of shape samples X channels.
% - labels (cell array): Cell array containing channel labels corresponding to the columns of the data matrix.
% - locs (numeric array): Matrix representing the locations of electrodes.
% - radius (numeric): Radius defining the neighborhood for Laplacian re-referencing.
%
% Returns:
% - out_values (numeric array): Laplacian-referenced EEG data matrix.
% - laplacian_labels (cell array): Channel labels for the Laplacian-referenced data.
%
% Example:
% values = randn(100, 5);  % Replace with your actual EEG data
% labels = {'Fp1', 'Fp2', 'C3', 'C4', 'O1'};
% locs = randn(5, 3);  % Replace with your electrode locations
% radius = 20;  % Replace with your desired radius
% [out_values, laplacian_labels] = laplacian(values, labels, locs, radius);
%
p = inputParser;
addRequired(p, 'values', @isnumeric);
addRequired(p, 'labels', @iscell);
addRequired(p, 'locs', @isnumeric);
addRequired(p, 'radius', @isnumeric);
parse(p, values, labels, locs, radius);
values = p.Results.values;
labels = p.Results.labels;
locs = p.Results.locs;
radius = p.Results.radius;

nchs = size(values,2);
out_values = nan(size(values));
laplacian_labels = cell(nchs,1);
close_chs = cell(nchs,1);
% do pseudo-laplacian if no locs info available
nan_elecs = any(isnan(locs),2);
[pseudo_values, pseudo_labels] = pseudo_laplacian(values,labels);
out_values(:,nan_elecs) = pseudo_values(:,nan_elecs);
laplacian_labels(nan_elecs) = pseudo_labels(nan_elecs);
% calculate distance
D = pdist2(locs,locs); 
close = D < radius;
close(logical(eye(size(close)))) = 0;

for i = 1:nchs
    if ~nan_elecs(i)
        close_elecs = close(i,:);
        if ~isempty(find(close_elecs))
            out_values(:,i) = values(:,i) - mean(values(:,close_elecs),2,'omitnan');
            laplacian_labels(i) = labels(i);
            close_chs{i} = find(close(i,:));
        else
            out_values(:,i) = pseudo_values(:,i);
            laplacian_labels(i) = pseudo_labels(i);
            close_chs{i} = 0;
        end
    else
        close_chs{i} = 0;
    end
end

end